src/core/providers/ogr/qgsogrprovider.cpp

```diff
@@ -1875,7 +1875,14 @@ bool QgsOgrProvider::addFeatures( QgsFeatureList &flist, Flags flags )
     returnvalue = false;
   }
 
-  recalculateFeatureCount();
+  if ( mFeaturesCounted != QgsVectorDataProvider::Uncounted &&
+       mFeaturesCounted != QgsVectorDataProvider::UnknownCount )
+  {
+    if ( returnvalue )
+      mFeaturesCounted += flist.size();
+    else
+      recalculateFeatureCount();
```

Comment:
What about just resetting the feature count to "uncounted" here, so that we defer the counting until it's actually needed?

Reply:
@nyalldawson QgsOgrProvider::featureCount() just returns mFeaturesCounted . I don't want to rewrite all the provider :-) Calling recalculeFeatureCount() here isn't really an issue as it is an error code path.

Reply:
Fair enough - I thought we already had logic here around deferred feature counting!